functions/go/apply-setters/main.go

```diff
@@ -25,64 +26,16 @@ func main() {
 		return nil
 	})
 
-	cmd.Long = usage()
+	cmd.Short = generated.ApplySettersShort
+	cmd.Long = generated.ApplySettersLong
+	cmd.Example = generated.ApplySettersExamples
+
 	if err := cmd.Execute(); err != nil {
 		fmt.Println(err)
 		os.Exit(1)
 	}
 }
 
-func usage() string {
```

Comment:
I'd assume something like

```
usage() {
  return readFromMarkdown(pathToMarkdown);
}
```

In the current formulation, we are not providing this functionality to our SDK users without some pre-processing step. We should keep in mind eventually we want parity between languages as well, so it cannot be specific to Go. Here's usage template used in TS SDK:

https://github.com/GoogleContainerTools/kpt-functions-sdk/blob/master/ts/create-kpt-functions/templates/func.mustache

Reply:
We can implement `make generate` for ts functions as well so that mdtogo generates usage strings in .ts file which can be used by ts files. As the `pathToMarkdown` is not shipped with binary, we need preprocessing to generate language strings from README doc.

Reply:
Two things here:
1. Ideally, we should provide such functionality to function developers automatically instead of asking them to set this up manually.
2. If possible, it's better to not have a preprocessing step at all (with the complexity of setting up make scripts, etc). Instead, can we just provide library utilities that read the `md` at runtime?

It's fine for us to set up preprocessing steps when we develop the kpt CLI (using Go-specific tooling), but asking function developers (in every supported language) to go through this process requires more thought.

Reply:
Parsing can't happen runtime. Binary/image can only include go files and not .md files. So we need a pre-processing step to convert README.md to go/ts files and ship them with binary.

Here are the steps for a function development.
1. Function author clones `kpt-functions-catalog` repo and `cd` into functions/go(or)ts folder.
2. Function author develops a function.
3. Author writes README.md
4. Author runs `make generate` to convert equivalent go/ts files for README.md

This also makes sure that the cli docs are consistent with Github docs. I know there are some additional steps here but let me know if you have a better idea to simplify the process.

Reply:
Functions can be written outside of the the catalog repo. We have a TS SDK that enables users to create an NPM package from scratch:
https://googlecontainertools.github.io/kpt/guides/producer/functions/ts/develop/

Asking them to fork the catalog repo to inherit some Make targets is a strictly worse UX.
An acceptable solution is to extend the SDK (npm scripts) to generate code from readme, but that functionality needs to be part of the SDK. Otherwise what's the point of having an SDK anyways?

We don't have a proper Go SDK today (just a library), but eventually we need it as well.

I don't feel strongly about pre-generation vs runtime, but the the latter is technically possible. Go binary can read a md file included in the container image. 

@mengqiy Please have tracking issues for the AI(s) here.

Reply:
https://github.com/GoogleContainerTools/kpt/issues/1642 and https://github.com/GoogleContainerTools/kpt/issues/1641